component/src/main/java/org/wso2/extension/siddhi/execution/streamingml/clustering/kmeans/KMeansIncrementalSPExtension.java

```diff
@@ -0,0 +1,295 @@
+                        type = {DataType.DOUBLE, DataType.FLOAT, DataType.INT, DataType.LONG}
+                )
+
+        },
+        returnAttributes = {
+                @ReturnAttribute(
+                        name = "euclideanDistanceToClosestCentroid",
+                        description = "Represents the Euclidean distance between the current data point and the " +
+                                "closest centroid.",
+                        type = {DataType.DOUBLE}
+                ),
+                @ReturnAttribute(
+                        name = "closestCentroidCoordinate",
+                        description = "This is a variable length attribute. Depending on the dimensionality(D) " +
+                                "we will return closestCentroidCoordinate1, closestCentroidCoordinate2,... " +
+                                "closestCentroidCoordinateD which are the d dimensional coordinates of the closest " +
+                                "centroid from the model to the current event. This is the prediction result and " +
+                                "this represents the cluster to which the current event belongs to.",
+                        type = {DataType.DOUBLE}
+                )
+        },
+        examples = {
+                @Example(
+                        syntax = "define stream InputStream (modelFeature1 double, modelFeature2 double);" +
+                                "from InputStream#streamingML:KMeansIncremental(modelName, numberOfClusters, " +
+                                "modelFeature1, modelFeature2)\"\n" +
+                                "select modelFeature1, modelFeature2, euclideanDistanceToClosestCentroid, " +
+                                "closestCentroidCoordinate1, closestCentroidCoordinate2\"\n" +
+                                "insert into OutputStream",
+                        description = "modelName='model1', numberOfClusters=2, This will select first two " +
+                                "distinct values in " +
+                                "events as initial model and predict and update the model for every event"
+                ),
+        }
+)
+public class KMeansIncrementalSPExtension extends StreamProcessor {
+    private double decayRate;
+    private int coordinateStartIndex;
+    private LinkedList<DataPoint> dataPointsArray;
+    private double[] coordinateValuesOfCurrentDataPoint;
+    private boolean isModelInitialTrained;
+    private Clusterer clusterer;
+    private int dimensionality;
+    private String modelName;
+    private static final Logger logger = Logger.getLogger(KMeansIncrementalSPExtension.class.getName());
+
+    @Override
+    protected List<Attribute> init(AbstractDefinition abstractDefinition,
+                                   ExpressionExecutor[] attributeExpressionExecutors,
+                                   ConfigReader configReader, SiddhiAppContext siddhiAppContext) {
+        dataPointsArray = new LinkedList<>();
+
```

Comment:
@niruhan shall we remove these unnecessary spaces?

Reply:
Fixed in https://github.com/siddhi-io/siddhi-execution-streamingml/pull/76